libraries/state_history/include/eosio/state_history/rocksdb_receiver.hpp

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <eosio/state_history/serialization.hpp>
+#include <eosio/chain/backing_store/db_combined.hpp>
+
+namespace eosio {
+namespace state_history {
+
+/*
+ *   Receiver of rocksdb data for SHiP delta generation.
+ *
+ *   It has 2 modes of operation:
+ *
+ *   1. When operating on the "single entry" mode, it will be used when working with single key entries, because
+ *   some of these objects (primary and secondary indexes) require information about the table they belong to, then for
+ *   these objects we will receive 2 calls, the first one related to the table_id_object_view (but in this case this object
+ *   does not contain proper values for payer so it can't be used to create its delta) associated with them and
+ *   the second call for the object itself. For that reason there is logic also associated to detect when we want
+ *   to store the table_id_object_view itself (which is when we receive only 1 call for this and the next call is not
+ *   any of the primary or secondary indexes).
+ *
+ *   2. When operating on the not sigle entry mode, then we store the deltas of the objects as we receive them.
+ *
+ */
+class rocksdb_receiver {
+public:
+   rocksdb_receiver(std::vector<table_delta> &deltas, const chainbase::database& db, bool single_entry = false) : deltas(deltas), db_(db), single_entry(single_entry) {};
+
+   void add_row(const chain::kv_object_view& row) {
+      if(single_entry) {
+         maybe_process_table();
+      }
+
+      deltas[get_delta_index("key_value")].rows.obj.emplace_back(present, fc::raw::pack(make_history_serial_wrapper(db_, row)));
+      table_ = {};
+   }
+
+   void add_row(const chain::backing_store::table_id_object_view& row) {
+      if(single_entry) {
```

Comment:
add a comment here to say that add_row can be called for the table row prior to processing the primary and secondary keys in that table, if it is not a single_entry_, or it can be called just to give context for a subsequent single primary or secondary key or it could be indicating that table key changed.  So when single_entry_ is set, we do not immediately report it as a delta till we see what comes next.  If it is followed by a row that requires the table context (i.e. calls add_row_with_table) then the table_id_object_view was just being provided as the context. If it is followed by any other row, then it was indicating a change in that table_id_object_view, and it needs to be added to the delta.

Reply:
I added a comment explaining this.